lib/src/gotrue_client.dart

```diff
@@ -57,89 +67,192 @@ class GoTrueClient {
 
-    return response;
+    if (authResponse.session != null) {
+      _saveSession(authResponse.session!);
+      _notifyAllSubscribers(AuthChangeEvent.signedIn);
+    }
+    return authResponse;
   }
 
-  /// Log in an existing user, or login via a third-party provider.
-  Future<GotrueSessionResponse> signIn({
+  /// Log in an existing user via a third-party provider.
+  Future<OAuthResponse> getOAuthSignInUrl({
+    required Provider provider,
+    String? redirectTo,
+    String? scopes,
+    Map<String, String>? queryParams,
+  }) async {
+    _removeSession();
+    return _handleProviderSignIn(provider,
+        redirectTo: redirectTo, scopes: scopes, queryParams: queryParams);
+  }
+
+  /// Log in a user using magiclink or a one-time password (OTP).
+  ///
+  /// If the `{{ .ConfirmationURL }}` variable is specified in the email template, a magiclink will be sent.
+  ///
+  /// If the `{{ .Token }}` variable is specified in the email template, an OTP will be sent.
+  ///
+  /// If you're using phone sign-ins, only an OTP will be sent. You won't be able to send a magiclink for phone sign-ins.
+  ///
+  /// If [shouldCreateUser] is set to false, this method will not create a new user. Defaults to true.
+  ///
+  /// [emailRedirectTo] can be used to specify the redirect URL embedded in the email link
+  ///
+  /// [data] can be user to set the user's metadata, which maps to the `auth.users.user_metadata` column.
+  ///
+  /// [captchaToken] Verification token received when the user completes the captcha on the site.
+  Future<AuthResponse> signInWithOtp({
     String? email,
     String? phone,
-    String? password,
-    Provider? provider,
-    OpenIDConnectCredentials? oidc,
-    AuthOptions? options,
+    String? emailRedirectTo,
+    bool? shouldCreateUser,
+    Map<String, dynamic>? data,
+    String? captchaToken,
   }) async {
     _removeSession();
 
-    if (email != null && password == null) {
-      await api.sendMagicLinkEmail(email, options: options);
-      return GotrueSessionResponse();
-    }
-    if (email != null && password != null) {
-      return _handleEmailSignIn(email, password, options: options);
-    }
-    if (phone != null && password == null) {
-      await api.sendMobileOTP(phone);
-      return GotrueSessionResponse();
-    }
-    if (phone != null && password != null) {
-      return _handlePhoneSignIn(phone, password);
-    }
-    if (provider != null) {
-      return _handleProviderSignIn(provider, options);
+    if (email != null) {
+      await _fetch.request(
+        '$_url/otp',
+        RequestMethodType.post,
+        options: GotrueRequestOptions(
+          headers: _headers,
+          body: {
+            'email': email,
+            'data': data ?? {},
+            'create_user': shouldCreateUser ?? true,
+            'gotrue_meta_security': {'captcha_token': captchaToken},
+          },
+        ),
+      );
+      return AuthResponse();
```

Comment:
`singInWithOtp` never returns a `user` or `session`, can't this method be of type `Future<void>` then?

Reply:
Makes sense! Let me add that update and merge this PR 🎉